lib/paseto.d.ts

```diff
@@ -0,0 +1,205 @@
+         */
+        protocol(): P;
+
+        /**
+         * encode the raw key as b64url
+         */
+        encode(): string;
+
+        /**
+         * return the raw key buffer
+         */
+        raw(): Buffer;
+    }
+
+    /**
+     * private key for asymmetric cryptography
+     */
+    export class PrivateKey<P extends IProtocol> implements IPasetoKey<P> {
+        constructor(proto: P);
+        public inject(rkey: Buffer): Promise<void>;
+        public base64(skey: string): Promise<void>;
+        public hex(skey: string): Promise<void>;
+        public generate(): Promise<void>;
+        public protocol(): P;
+        public encode(): string;
+        public raw(): Buffer;
+
+        /**
+         * return the corresponding public key object
+         */
+        public public(): Promise<PublicKey<P>>
+    }
+
+    /**
+     * public key for asymmetric cryptography
+     */
+    export class PublicKey<P extends IProtocol> implements IPasetoKey<P> {
+        constructor(proto: P);
+        public inject(rkey: Buffer): Promise<void>;
+        public base64(skey: string): Promise<void>;
+        public hex(skey: string): Promise<void>;
+        public generate(): Promise<void>;
+        public protocol(): P;
+        public encode(): string;
+        public raw(): Buffer;
+    }
+
+    /**
+     * secret key for symmetric cryptography
+     */
+    export class SymmetricKey<P extends IProtocol> implements IPasetoKey<P> {
+        constructor(proto: P);
+        public inject(rkey: Buffer): Promise<void>;
+        public base64(skey: string): Promise<void>;
+        public hex(skey: string): Promise<void>;
+        public generate(): Promise<void>;
+        public protocol(): P;
+        public encode(): string;
+        public raw(): Buffer;
+    }
+
+    namespace PrivateKey {
+        /**
+         * shortcut for new PrivateKey(new V1())
+         */
+        export class V1 extends PrivateKey<Paseto.V1> {
+            constructor();
+        }
+
+        /**
+         * shortcut for new PrivateKey(new V2())
+         */
+        export class V2 extends PrivateKey<Paseto.V2> {
+            constructor();
+        }
+    }
+
+    namespace PublicKey {
+        /**
+         * shortcut for new PublicKey(new V1())
+         */
+        export class V1 extends PublicKey<Paseto.V1> {
+            constructor();
+        }
+
+        /**
+         * shortcut for new PublicKey(new V2())
+         */
+        export class V2 extends PublicKey<Paseto.V2> {
+            constructor();
+        }
+    }
+
+    namespace SymmetricKey {
+        /**
+         * shortcut for new SymmetricKey(new V1())
+         */
+        export class V1 extends SymmetricKey<Paseto.V1> {
+            constructor();
+        }
+
+        /**
+         * shortcut for new SymmetricKey(new V2())
+         */
+        export class V2 extends SymmetricKey<Paseto.V2> {
+            constructor();
+        }
+    }
+
+    interface IProtocol {
```

Comment:
Similar to the above, I'd like to see distinct interfaces for both protocol versions.

Reply:
IProtocolV1 and IProtocolV2 created.